Fix SingleChoiceInput accessibility issues

foxbunny opened the pull request:
- Replace SUI's Dropdown component with a custom Combobox component
- Implement a custom `<ind-combobox>` element used by Combobox
- Add `useNativeEvent` hook
- Tweak regform styles to handle the new controls

micsucmed commented:
Hi @foxbunny, could you please explain in more detail what accessibility issues you're trying to solve here? I think you meant usability. There is still an issue when the user types in an entry that is not part of the drop-down options; there is no feedback that this entry is not valid and will not be sent (Given that the field is not required). And even if the field is required, the feedback gotten is "This field is required", which can be confusing since the user might think the field input has been provided, but in reality the input is not valid.
![image](https://github.com/indico/indico/assets/53895656/11b24d6e-cd23-4eb1-b5c7-a60329060f89)

foxbunny commented:
> please explain in more detail what accessibility issues you're trying to solve here

The primary accessibility issues that are being solved is incompatibility of the SUI combobox with screen readers. In particular, not announcing the options, not being clear about what effect the typing has, etc.

> There is still an issue when the user types in an entry that is not part of the drop-down options

I've already pointed this out to @ThiefMaster. It's a validation issue, not really an accessibility issue.

tomasr8 commented:
Thanks for the PR!

> I've already pointed this out to @ThiefMaster. It's a validation issue, not really an accessibility issue.

Can we do something about it in this PR?


Here's just some small styling issues I noticed:

When expanded I'd disable the bottom border radius on the input field:
![image](https://github.com/indico/indico/assets/8739637/d67adc90-b75d-40ab-8970-1b230611c68f)

I'd align the options with text input as it was before (i.e. increase left padding for the options)
![image](https://github.com/indico/indico/assets/8739637/c55767db-4982-48a7-89c5-e753c9896a4e)

The list needs a higher z-index:
![image](https://github.com/indico/indico/assets/8739637/60df3ce5-9101-4bff-b0c7-18729f691e25)

From a usability point of view, I noticed that if I have two options, one uppercase and one lowercase e.g. `A` and `a`, when I type lowercase `a`, the `A` option is auto selected. That's not what I'd expect, especially since it changes the character I typed (from `a` to `A`)

foxbunny commented:
> Can we do something about it in this PR?

My understanding was that it wasn't something we can do, but I can check the code, sure.

> Here's just some small styling issues I noticed:

Gotcha. Thanks.

> From a usability point of view, I noticed that if I have two options, one uppercase and one lowercase e.g. A and a

Hmmm. I'm not sure we should support this case at all. See, for screenreaders, there's absolutely no difference between capital A and lowercase a. So even if I make the combobox do case-sensitive matching in case of conflicts, it won't help screenreader users one bit.

foxbunny commented:
> I can imagine regular users not being happy with the combobox changing their input.. If it doesn't hurt screenreader users I'd prefer case-sensitive matching

Do you really intend to have options that are identical except for the case? If you ask me, I'd forbid that during form creation.

foxbunny commented:
Anyway, back to the discussion on case-matching. Suppose we have the following items in the list:

- ACME
- Acme
- acme
- aCme
- aCMe

The most trivial thing we can do is match at the start of the string case-sensitively. For example, if we type "A", it matches "ACME". If we type "Ac", it matches "Acme" and not "ACME". 

This has some edge cases. For example, if we type "acM" it will match "ACME" and not "acme" or "aCme" even though the first character is lower-case.

How elaborate do you want this algorithm to be?

tomasr8 commented:
> How elaborate do you want this algorithm to be?

A more radical idea, but what if we just don't autocomplete? As in, we keep the behaviour of the current dropdown?
My issue is purely with the fact that the input is actively changing what I'm typing, but maybe others have a different opinion? @ThiefMaster @micsucmed 

On an unrelated note:
- In the current dropdown one can search for any part of the text e.g. if you type `cme` you'll see `acme` as an option. I would prefer not to lose this functionality as it makes the search more useful
- A quick way to unselect the currently-selected option would be nice. The SUI dropdown has this `x` icon on the right side, maybe we could add it as well?

foxbunny commented:
> In the current dropdown one can search for any part of the text e.g. if you type cme you'll see acme as an option. I would prefer not to lose this functionality as it makes the search more useful

I deliberately avoided using that approach because it causes usability issues such as matching something that isn't obviously what the user started off to type. (For instance in the country selector you start to type "Ser" it will give you "Montserrat" as the first option.) But anyway, I leave that up to you.

> A quick way to unselect the currently-selected option would be nice. The SUI dropdown has this x icon on the right side, maybe we could add it as well?

Sure, we can add that.

foxbunny commented:
Updated the PR.

Fix styling issues, and implemented two strategies for handling typing based on the `aria-autocomplete` attribute, so now you can do both autocomplete and search (documented). Also added the clear button.

foxbunny commented:
Rebased on master, and also fixed the validation logic so it correctly traps invalid choices.

foxbunny commented:
@tomasr8 Reminder, just in case.

tomasr8 commented:
I noticed some differences from the old dropdown. Some of these might be nice to keep for the new one as well:

 - no results message:

![obrazek](https://github.com/indico/indico/assets/8739637/a59c3362-7870-45c8-8c06-2b6e57f7c6bd)
 
- I'd keep the same `x` icon:

![obrazek](https://github.com/indico/indico/assets/8739637/c1da0f78-401a-4a9c-895b-69521f0cd06b)
vs
![obrazek](https://github.com/indico/indico/assets/8739637/1a75e8d4-37f4-4158-96c7-a72c254702d1)

- When you have extra slots the dropdown to select the amount is much wider now (and the combobox changes width):

![obrazek](https://github.com/indico/indico/assets/8739637/3f2b5ed4-ebac-4d99-80c5-ec9566ce7842)
vs
![obrazek](https://github.com/indico/indico/assets/8739637/5d7d5c49-1d8d-4c7c-9844-7762b3a6bd56)

- The hover bg color is the same as the one for the price now so it blends together:

![obrazek](https://github.com/indico/indico/assets/8739637/d63565e0-0be4-4872-a49e-c1d85587d5e3)
vs
![obrazek](https://github.com/indico/indico/assets/8739637/441ec21b-38c9-40b0-b15d-b89932a984b1)

I'd also increase the height of the listbox a bit, currently it feels a bit too small

foxbunny commented:
> When you have extra slots the dropdown to select the amount is much wider now (and the combobox changes width)

Do we even need the amount drop-down to be a combobox? I can replace it with a plain select list.

foxbunny commented:
> no results message:

It would take a lot of additional code to get this feature in. I think there isn't such a dramatic difference between saying there are no matches and showing no matches from the UX standpoint, and therefore the additional work, maintenance burden, and complexity aren't justified. Of course, if you insist that this feature is crucial, I can still add it.

tomasr8 commented:
> > no results message:
> 
> It would take a lot of additional code to get this feature in. I think there isn't such a dramatic difference between saying there are no matches and showing no matches from the UX standpoint, and therefore the additional work, maintenance burden, and complexity aren't justified. Of course, if you insist that this feature is crucial, I can still add it.

Ok if it'd require a lot of extra work we can leave it out

> Do we even need the amount drop-down to be a combobox? I can replace it with a plain select list.

I think so :)

foxbunny commented:
> I think so :)

Ok. I've pushed a fix for the combobox widths. Right now, these controls use different implementations (number of extra slots uses SUI dropdown). My plan is to replace the other ones once this PR is merged so we don't do too much in a single PR. Meanwhile you can adapt this to the time picker, for instance. Let me know if that works for you.

tomasr8 commented:
> Ok. I've pushed a fix for the combobox widths. Right now, these controls use different implementations (number of extra slots uses SUI dropdown). My plan is to replace the other ones once this PR is merged so we don't do too much in a single PR. Meanwhile you can adapt this to the time picker, for instance. Let me know if that works for you.

As long as the look is consistent across the different dropdowns it should be fine

ThiefMaster commented on indico/web/client/js/react/components/Combobox.jsx:
```diff
@@ -0,0 +1,67 @@
+          );
+        })}
+      </ul>
+      <button type="button" value="clear">
```
```suggestion
      <button type="button" value="clear" disabled={inputProps.disabled}>
```

ThiefMaster replied:
I already applied this since it was a very simple fix and it worked fine when testing it locally.

ThiefMaster commented on indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:
```diff
@@ -75,53 +76,47 @@ function SingleChoiceDropdown({
-          placeholder={Translate.string('Choose an option')}
+          onChange={handleChange}
           options={options}
-          disabled={disabled}
```
I already added this since it worked fine when testing it locally.

ThiefMaster commented on indico/web/client/js/react/components/Combobox.jsx:
```diff
@@ -0,0 +1,67 @@
+
+  return (
+    <ind-combobox>
+      <input ref={inputRef} {...uncontrolledInputProps} type="text" role="combobox" />
```
This needs `autoComplete="off"` to avoid ugly cases such as this one:

![image](https://github.com/indico/indico/assets/179599/63d4709a-a500-4ed0-b892-8eed837cad5c)

ThiefMaster commented:
@foxbunny I rebased this PR to the latest master and applied a few minor fixes for very straightforward things I encountered while testing.

I think finishing this PR could actually get higher priority on your side than the "datepicker for accommodation field" suggestion I made earlier today, since this one feels almost ready for a merge into master, and that way we'll have one less pending PR plus the advantage of being able to use the new Combobox (or should we call it `ComboBox`? ;)) elsewhere!

ThiefMaster commented on indico/web/client/js/react/components/Combobox.jsx:
```diff
@@ -0,0 +1,67 @@
+          );
+        })}
+      </ul>
+      <button type="button" value="clear" disabled={inputProps.disabled}>
```
Is clicking this button supposed to open the dropdown? It feels strange because when I clear the field it typically means I want to to be empty, because in case I want to change the value I can simply open it again instead of clearing it first.

ThiefMaster commented on indico/web/client/js/react/components/Combobox.jsx:
```diff
@@ -0,0 +1,67 @@
+          );
+        })}
+      </ul>
+      <button type="button" value="clear" disabled={inputProps.disabled}>
```
This button should not be displayed, if the field is `required`.

foxbunny replied:
While fixing this, I noticed an interesting feature we got for free. If we set the input to `readonly`, then we get something that works like a select list. 😊

foxbunny commented on indico/web/client/js/custom_elements/ind_combobox.js:
```diff
@@ -0,0 +1,288 @@
+// This file is part of Indico.
```
If it cannot be selected, maybe it makes more sense to simply remove it from the list? Or is the information that it's not longer available that important?

ThiefMaster replied:
I think it's important to keep showing it. When buying tickets for something they usually also show sold-out options and don't simply make them disappear.

foxbunny replied:
Gotcha.

foxbunny commented:
@ThiefMaster I think I've taken care of the issues discussed here. I've also a minor accessibility issue with the price tag label not being associated with the extra slots field, so fixed that as well. Let me know if you spot anything else.

tomasr8 commented:
Thanks! I have some more style nitpicks ;)

- For disabled choices, I'd use the same lighter font color for the price as well (as it was previously).
Previous:
![image](https://github.com/indico/indico/assets/8739637/65c86b56-6a3d-42cb-a9e7-7285a96fe1e0)
Current:
![image](https://github.com/indico/indico/assets/8739637/ec0fa083-404d-4ba9-ac7c-c85b3bfe2143)

- The right alignment of the price label is a bit weird now (compare the first and second choice)
![image](https://github.com/indico/indico/assets/8739637/ec0fa083-404d-4ba9-ac7c-c85b3bfe2143)

- I think the default height could be a bit larger. Currently it fits only 4 choices, which makes it feel a bit cramped

- I might be wrong but I don't think the clearable `X` button should be visible when the field is required:
![image](https://github.com/indico/indico/assets/8739637/b1a63cf0-e86c-4498-b4bf-dc92e89bf1f1)

foxbunny commented:
@tomasr8 Fixed.

tomasr8 commented:
Thanks again!

> I think the default height could be a bit larger. Currently it fits only 4 choices, which makes it feel a bit cramped

Much better now! Do you think we could increase it up to 8 to match the old dropdown?

Would it be possible to not have the line wrapping on the price label?
Current:
![image](https://github.com/indico/indico/assets/8739637/a4abf896-a7e4-4fc4-b3c4-d6ee2bf2d5f8)
Previous:
![image](https://github.com/indico/indico/assets/8739637/ea0f6297-9e30-472e-b347-58f7606a5f17)

foxbunny commented:
> Much better now! Do you think we could increase it up to 8 to match the old dropdown?

6~7 items is considered maximum for most humans for optimal operation.

> Would it be possible to not have the line wrapping on the price label?

Sure

ThiefMaster commented on indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:
```diff
@@ -68,73 +79,85 @@ function SingleChoiceDropdown({
-        value={value[selectedChoice.id]}
-        onChange={(_evt, data) => onChange({[selectedChoice.id]: data.value})}
-        fluid
+        value={String(selectedSeats)}
```
When changing between two options there's a bug when 'extra slots' are enabled:

- The combobox value is not cleared, so if I switch from a max 3 slots choice (w/ 3 selected) to one that allows max 2 slots, the combobox still shows "3"
- Regardless of the choice's validity, when changing the option the combobox keeps the displayed value, which is now different from the submitted value. So If I change the choice like I mentioned in my previous point, it looks like I'm submitting the choice with 3 slots, but in reality I submit it with 1 slot.

foxbunny replied:
Will check

ThiefMaster replied:
Likely related: When you use the `x` icon to clear the selected choice, the slots dropdown does not disappear.

ThiefMaster commented on indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:
```diff
@@ -68,73 +79,85 @@ function SingleChoiceDropdown({
 
+  const handleChange = evt => {
+    if (!evt.target.value) {
+      onChange({'': -1});
```
Setting a value of `{"": -1}` seems wrong. It should be `{}` when nothing is selected. This is especially important if the field is not required, because in that case the "empty" value can actually get submitted.

ThiefMaster commented on indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:
```diff
@@ -68,73 +79,85 @@ function SingleChoiceDropdown({
+      onChange({'': -1});
+      return;
+    }
+    const selected = choices.find(c => c.caption === evt.target.value);
```
I don't like this lookup based on the caption... it's basically a roundtrip from unique id to (not guaranteed-to-be-unique) caption to unique id.

Can't we use the option value as the value of the ind-combobox? This would also match the standard HTML `<select>` `<option>` behavior more closely.

And this is functionality we'll certainly need to use this combobox elsewhere since there we typically don't want to have lots of boilerplate code around it to implement option values...

foxbunny replied:
> I don't like this lookup based on the caption... it's basically a roundtrip from unique id to (not guaranteed-to-be-unique) caption to unique id.

That lookup has to happen somewhere anyway because this is *not* a select box. Then we again have this interpretation vs physical reality situation we had with "invalid" property.

The difference between combobox and a select box is that a combobox is fully editable, which means user can enter anything they want. If we don't want then, then we want to implement a different widget: searchable select list. That would mean a non-editable label coupled with a text field whose sole purpose is to filter the list.

As for uniqueness of the captions, if you ask me, that absolutely must be guaranteed in the management area, as having two identical captions with different values simply adds confusion for the user. How do you pick one if they are the same??

ThiefMaster replied:
> That lookup has to happen somewhere anyway because this is not a select box

Does it? In the end you create options from a list of options that typically have a label and a value. Couldn't the value of the selected option be stored in a member variable of the ind-combo-box instance or a hidden field? That way you'd never need to derive the value from the caption.

> As for uniqueness of the captions, if you ask me, that absolutely must be guaranteed in the management area, as having two identical captions with different values simply adds confusion for the user. How do you pick one if they are the same??
I agree that we typically do not want non-unique captions. But there are some cases - albeit very rare - where it could happen: Someone selects an option, then that option is removed by a manager (existing registrations keep it!), and later another option with the same caption is added back.

The dropdown will then show the old option (IIRC with an indicator next to the field that you may not be able to select it again if you deselect it and save), but also any new options since those would be valid choices now as well.

> The difference between combobox and a select box is that a combobox is fully editable, which means user can enter anything they want.

Maybe there was a misunderstanding, but that functionality never existed in Indico. At least not in the registration form (we use the SUI Dropdown with `allowAdditions` in some places, typically for an editable list of things like file extensions). The dropdown for the Single Choice registration form field was always supposed to be a (searchable) select list!

foxbunny replied:
> Does it? In the end you create options from a list of options that typically have a label and a value.

The intended use case for combobox is basically exactly the same as using a `<datalist>`. It simply provides candidates for the text input. This is why it's allowed for the input to also perform the search. From the perspective of the end user, they are entering a value into the input. 

Select list is not editable because it does not allow arbitrary input. Since it is not editable, there is no expectation that an arbitrary value is allowed.

> Couldn't the value of the selected option be stored in a member variable of the ind-combo-box instance or a hidden field? That way you'd never need to derive the value from the caption.

That's technically doable (and that's similar to what SUI does), but it breaks the user's intuition about how the field works, and this is especially true for people who cannot see the field once they've left it. In order to correct the perception of what's to be expected of the field, the field itself cannot be editable. Therefore, a separate control is provided to facilitate searching and it's part of the popup listbox, not the form control.

> The dropdown will then show the old option (IIRC with an indicator next to the field that you may not be able to select it again if you deselect it and save), but also any new options since those would be valid choices now as well.

In that case, you can't use a combobox. It will be semantically very confused. I recommend using a select list with separate search in the listbox.

ThiefMaster replied:
Then I think there was indeed some misunderstanding somewhere that made you go for a combobox :/

Anyway, I think we can still keep it around in case it becomes useful at some point elsewhere in Indico?

foxbunny replied:
Yeah, if you need an actual text input that supports completion candidates, then combobox will work as expected. I'll add a searchable select list to this PR and rewire the SingleSelectInput to use that instead of the combobox and keep the combobox custom element around.

foxbunny replied:
Before I do the PR, I'll implement a prototype in a sandbox, tho, and then we can take it from there.

foxbunny replied:
@ThiefMaster Check this out:

https://codepen.io/hayavuk/pen/mdYLYEL?editors=0110

It's still not 100% done yet, but it shows the behavior.

In general, it behaves more or less like the native select, with the differences being:

- It automatically expands the list when you press the arrow keys. I think that behavior is ok, but it's actually a hack to work around the issue of changes in value not being announced to screen readers when the list is closed.
- It has a search box.
- It can display arbitrary elements other than just the value label.

tomasr8 replied:
Works nice! I just got a question regarding the look of the element - are you planning to leave the search input inside the listbox? I'd like to keep the new select list similar to the combobox for uniformity and have the search input on top. What do you think?

foxbunny replied:
That doesn't make sense to me. How would you differentiate between selected value and random input?

tomasr8 replied:
We could just remove it when the field loses focus, like SUI does: https://react.semantic-ui.com/modules/dropdown/#types-search-selection

foxbunny replied:
That's the whole problem. You're not supposed to edit the field *after* it loses focus unless it's clear that the field is temporary (e.g., a filter for the list). This is where the whole thing about the original combobox started. SUI's combobox behavior is bad, I tried to fix it, and then it turned out it's bad because it's trying to present a searchable select list as a combobox.

foxbunny replied:
Also, I've updated the example to include a way to clear the field. You can add the `required` attribute to the `<x-select>` just like you would to any other control to see its behavior when the field is required.

foxbunny replied:
@tomasr8 I re-read your original comment, and I think this is basically what you had in mind:

https://codepen.io/hayavuk/pen/jOoomaN?editors=0100

This actually doesn't affect the separation between selected value caption and the search input. It simply visually places the input "on top" of the control caption so that it appears as if the caption is editable. I used the placeholder instead of populating the value as a work-around for the issues I mentioned above.

tomasr8 replied:
Just for reference, here's another example from MUI: https://mui.com/material-ui/react-autocomplete/#combo-box

foxbunny replied:
Not sure what you mean by "add an extra value". Can you elaborate?

ThiefMaster commented on indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:
```diff
@@ -68,73 +79,85 @@ function SingleChoiceDropdown({
   }));
 
+  const handleChange = evt => {
+    if (!evt.target.value) {
```
There's a bug when the field is not required and you type some nonsense in it: In that case there's no validation error, because it'll make no difference here if you explicitly cleared the field or wrote nonsense in it.

I think we need to add something similar here like what I did in the `FinalDatePicker`...

foxbunny replied:
Tbh, this is starting to sound like you don't want a combobox at all, but a searchable select list. Combobox really should be validated like any text field and then it should work as expected.

ThiefMaster replied:
Yeah, this was never meant to be a combobox... (more details in my [other comment](https://github.com/indico/indico/pull/6310#discussion_r1646500737))

foxbunny commented:
@ThiefMaster @tomasr8 The select list has been implemented. I tweaked the behavior a bit and made it automatically select the last option that exactly matches the keyword (case-insensitive). Other than that, it behaves pretty much as in the prototype.

ThiefMaster reviewed (commented):
I tested it and it works fine AFAICT :)
Just a few small comments as usual.

foxbunny commented on indico/web/client/js/react/components/Select.jsx:
```diff
@@ -0,0 +1,93 @@
+  return (
+    <ind-select ref={indSelectRef} value={value} {...inputProps}>
+      <div className="caption" data-caption>
+        <Translate>Select a choice</Translate>
```
This text isn't a placeholder though. It's an important cue that lets you tell this widget apart from a plain input. Therefore, I think it should not fail the basic contrast check for a11y. Tbh, I think placeholders shouldn't either (e.g., failing contrast check on a placeholder means important information such as date format is *imperceptible* to users with poor vision), but it's accepted as "industry standard".

ThiefMaster replied:
TBH I find it a bit hard to distinguish between "the field has a value" and "the field has no value" right now...

foxbunny replied:
The native select works exactly like that:

<img width="147" alt="image" src="https://github.com/user-attachments/assets/e909ad19-16ec-4e81-b0f1-e4ca7897e620">

foxbunny commented on indico/web/client/styles/partials/_sui_debt.scss:
```diff
@@ -18,3 +20,60 @@
+}
+
+@mixin border-radius {
+  border-radius: 0.28571429rem;
```
Everything in that module needs to eventually be phased out and replaced with the design system. I wouldn't worry about it too much.

foxbunny replied:
If those values are used outside SUI, such declarations should be moved into this module. Within this module, we basically shouldn't care how dirty something is. It's not intended for long-term maintenance.

foxbunny commented on indico/web/client/js/custom_elements/ind_select.scss:
```diff
@@ -0,0 +1,192 @@
+    text-overflow: ellipsis;
+  }
+
+  &:focus .caption {
```
I think I fixed this one. It should've been `:focus-visible` instead of `:focus`.

foxbunny replied:
I'll look into why the other field receives focus, though. Cause it shouldn't. Maybe it's a Firefox bug or something.

foxbunny replied:
I don't see anything obvious in the code that would cause such behavior, so I'd say it's definitely a Firefox bug. Chrome doesn't have this issue. It doesn't affect keyboard users, and we took care of the visual glitch for mouse users, so I'm inclined to say it's good enough as it is.

ThiefMaster replied:
works all fine now :)

ThiefMaster commented on indico/web/client/js/react/components/Select.jsx:
```diff
@@ -0,0 +1,100 @@
+// This file is part of Indico.
```
I noticed this really weird behavior in Firefox where it randomly selects text when repeatedly opening and closing the dropdown... could be a Firefox bug though, so I would not spend *too* much time on it...

![weird-ff-selection](https://github.com/user-attachments/assets/273de644-91b1-48b7-801d-a25d259dc325)

ThiefMaster replied:
Actually it seems to happen when moving the mouse cursor just the slightest bit while clicking.

foxbunny replied:
Well, that's a common issue with any custom widget. We can mask it with CSS so at least it's not visible, but not sure if selection can be outright prevented or if it's even worth it.

ThiefMaster replied:
I never noticed it with other widgets... Would deferring the closing of the dialog help, so the click event has already been fully handled by the time it closes?

foxbunny replied:
> I never noticed it with other widgets...

Happens quite often actually.

> Would deferring the closing of the dialog help, so the click event has already been fully handled by the time it closes?

Will see what I can do.

tomasr8 commented:
I'm getting this weird exception when I try to drag&drop the widget in the form designer (`manage/registration/XXX/form/`):

![image](https://github.com/user-attachments/assets/169b613e-5cae-4366-9b84-f7ae68e14aa3)

foxbunny commented:
Ah, damn. My bad. I totally overlooked the drag & drop. It will probably affect the date picker as well. I'll do a separate PR for that.

foxbunny commented:
Ok, please ignore the last fix. I'm going to try and come up with a more systematic way to address these temporary disconnects so that it can be applied to all custom elements. When I'm done, I'll first do a PR to address that in the datepicker, and then I'll use that fix in this PR as well.

foxbunny commented:
This is the fix that I propose for this issue. 

https://github.com/indico/indico/pull/6447

foxbunny commented:
Updated the PR to include the base class in combobox and select.

ThiefMaster commented on CHANGES.rst:
```diff
@@ -120,6 +120,7 @@ Accessibility
 ^^^^^^^^^^^^^
 
 - Improve registration form date picker accessibility (:pr:`6371`, thanks :user:`foxbunny`)
+- Improve registration form single choice input accessibility (:pr:`6310`, thanks :user:`foxbunny`)
```
Please move the changelog entries to 3.3.4 (a11y section goes between bugfixes and internal changes, since there isn't one for 3.3.4 yet).

ThiefMaster commented on indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:
```diff
@@ -370,6 +398,27 @@ export default function SingleChoiceInput({
+
+    for (const key in value) {
+      if (!optionSet.has(key)) {
+        return Translate.string('Please select one of the provided choices');
```
Can this still happen? I think it might be a leftover from when it was a ComboBox

ThiefMaster commented on indico/web/client/js/custom_elements/ind_select.js:
```diff
@@ -0,0 +1,323 @@
+        }
+        toggleListbox(false);
+        indSelect._closedViaFocusout = true;
+        setTimeout(function() {
```
```suggestion
        setTimeout(() => {
```

ThiefMaster commented on indico/web/client/js/custom_elements/ind_select.js:
```diff
@@ -0,0 +1,323 @@
+        if (indSelect.required) {
+          internals.setValidity(
+            {valueMissing: !indSelect.value},
+            'Please select an item in the list.'
```
`Translate.string(...)`

ThiefMaster commented on indico/web/client/js/custom_elements/ind_select.js:
```diff
@@ -0,0 +1,323 @@
+      }
+
+      function selectNextOption() {
+        selectOptionWithIndex(function(currentIndex, optionCount) {
+          return (currentIndex + 1) % optionCount;
+        });
```
```suggestion
        selectOptionWithIndex((currentIndex, optionCount) => (currentIndex + 1) % optionCount);
```

ThiefMaster commented on indico/web/client/js/custom_elements/ind_select.js:
```diff
@@ -0,0 +1,323 @@
+      }
+
+      function selectPreviousOption() {
+        selectOptionWithIndex(function(currentIndex, optionCount) {
+          if (currentIndex <= 0) {
+            return optionCount - 1;
+          }
+          return currentIndex - 1;
+        });
```
```suggestion
        selectOptionWithIndex((currentIndex, optionCount) =>
          currentIndex <= 0 ? optionCount - 1 : currentIndex - 1
        );
```
